packages/agent/src/polling/index.ts

```diff
@@ -1,53 +1,159 @@
+    hasProperty(value, 'body') &&
+    value.body !== null &&
+    typeof value.body === 'object' &&
+    hasProperty(value.body, 'content') &&
+    value.body.content !== null &&
+    typeof value.body.content === 'object' &&
+    hasProperty(value.body.content, 'request_type') &&
+    value.body.content.request_type === ReadRequestType.ReadState &&
+    hasProperty(value.body.content, 'ingress_expiry') &&
+    typeof value.body.content.ingress_expiry === 'object' &&
+    value.body.content.ingress_expiry !== null &&
+    hasProperty(value.body.content.ingress_expiry, 'toCBOR') &&
+    typeof value.body.content.ingress_expiry.toCBOR === 'function' &&
+    hasProperty(value.body.content.ingress_expiry, 'toHash') &&
+    typeof value.body.content.ingress_expiry.toHash === 'function'
+  );
+}
+
 /**
  * Polls the IC to check the status of the given request then
  * returns the response bytes once the request has been processed.
  * @param agent The agent to use to poll read_state.
  * @param canisterId The effective canister ID.
  * @param requestId The Request ID to poll status for.
- * @param strategy A polling strategy.
- * @param request Request for the repeated readState call.
- * @param blsVerify - optional replacement function that verifies the BLS signature of a certificate.
+ * @param options - polling options to control behavior
  */
 export async function pollForResponse(
   agent: Agent,
   canisterId: Principal,
   requestId: RequestId,
-  strategy: PollStrategy = defaultStrategy(),
-  request?: unknown,
-  blsVerify?: CreateCertificateOptions['blsVerify'],
+  options: PollingOptions = {},
 ): Promise<{
   certificate: Certificate;
   reply: ArrayBuffer;
 }> {
-  const path = [new TextEncoder().encode('request_status'), requestId];
-  const currentRequest = request ?? (await agent.createReadStateRequest?.({ paths: [path] }));
+  const encode = (str: string) => bufFromBufLike(new TextEncoder().encode(str));
+  const path = [encode('request_status'), requestId];
+
+  // Determine if we should reuse the read state request or create a new one
+  // based on the options provided.
+  async function constructRequest(paths: ArrayBuffer[][]): Promise<ReadStateRequest> {
+    if (options.request && isSignedReadStateRequestWithExpiry(options.request)) {
+      return options.request;
+    }
+    const request = await agent.createReadStateRequest?.(
+      {
+        paths,
+      },
+      undefined,
+    );
+    if (!isSignedReadStateRequestWithExpiry(request)) {
+      throw new CreateReadStateRequestError('Invalid read state request', request);
+    }
+    return request;
+  }
+
+  let state;
+  let currentRequest: ReadStateRequest | undefined = undefined;
+  const preSignReadStateRequest = options.preSignReadStateRequest ?? false;
+  if (preSignReadStateRequest) {
+    // If preSignReadStateRequest is true, we need to create a new request
+    currentRequest = options.request ?? (await constructRequest([path]));
```

Comment:
Since we already check for the `options.request` in the `constructRequest` function, shouldn't we change this line to the following? Otherwise, we skip the `isSignedReadStateRequestWithExpiry` check.

```suggestion
    currentRequest = await constructRequest([path]);
```